File: GenerateParenthesis.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> vals = new ArrayList<String>();
        ParenthesisRecurs(0, 0, vals, n, "");
        return vals;
    }

    public void ParenthesisRecurs(int open, int closed, 
    List<String> list, int n, String parenthesis){
        if(open == closed && closed == n){
            list.add(parenthesis);
            return;
        }
        if(open < n){
            ParenthesisRecurs(open+1, closed, list, n, parenthesis + "(");
        }
        if(closed < open){
            ParenthesisRecurs(open, closed+1, list, n, parenthesis + ")");
        }
    }
}

File: GenerateParenthesis.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        stack = []
        res = []

        def makeparenthesis(open, closed):
            if open == closed == n:
                res.append("".join(stack))
                return
            if open < n:
                stack.append("(")
                makeparenthesis(open+1, closed)
                stack.pop()
            if closed < open:
                stack.append(")")
                makeparenthesis(open, closed+1)
                stack.pop()
        
        makeparenthesis(0, 0)
        return res
            
